composer/datasets/synthetic_lm.py

```diff
@@ -0,0 +1,231 @@
+        decoder=decoders.ByteLevel(),
+        initial_alphabet=pre_tokenizers.ByteLevel.alphabet(),
+        special_tokens=[pad_token, "<endoftext>"],
+        pad_token=pad_token,
+        trainer_cls=tokenizers_trainer.BpeTrainer,
+        tokenizer_cls=GPT2Tokenizer,
+    )
+
+
+def generate_synthetic_tokenizer(tokenizer_family: str,
+                                 dataset: Optional[Dataset] = None,
+                                 vocab_size: int = 256) -> PreTrainedTokenizer:
+    """Generates a synthetic tokenizer based on a tokenizer family.
+
+    Args:
+        tokenizer_family (str): Which tokenizer family to emulate. One of ['gpt2', 'bert'].
+        dataset (Optional[datasets.Dataset]): Optionally, the dataset to train the tokenzier off of.
+            If not specified, a SyntheticHFDataset will be generated.
+        vocab_size (int): The size of the tokenizer vocabulary. Defaults to 256.
+    """
+
+    try:
+        from tokenizers import Tokenizer
+        from transformers import PreTrainedTokenizer
+    except ImportError as e:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without NLP support. To use NLP with Composer, run `pip install mosaicml[nlp]`
+            if using pip or `conda install -c conda-forge transformers tokenizers` if using Anaconda.""")) from e
+
+    # generate a synthetic dataset with reasonable defaults is none is provided
+    if dataset is None:
+        num_samples = 100
+        chars_per_sample = 128
+        column_names = ['sentence']
+        dataset = SyntheticHFDataset(num_samples=num_samples,
+                                     chars_per_sample=chars_per_sample,
+                                     column_names=column_names).generate_dataset()
+
+    # change a columnar dataset into a list
+    flattened_columns = [dataset[key] for key in dataset.column_names if key != 'idx']
+    # flatten the list of lists into a single list
+    flattened_dataset = []
+    for sublist in flattened_columns:
+        for item in sublist:
+            flattened_dataset.append(item)
+
+    if "bert" in tokenizer_family:
+        tokenizer_params = _generate_bert_tokenizer_params(flattened_dataset)
+    elif "gpt2" in tokenizer_family:
+        tokenizer_params = _generate_gpt2_tokenizer_params()
+    else:
+        raise ValueError(f"Synthetic tokenizers for tokenizer family {tokenizer_family} are currently unsupported.")
+
+    tokenizer = Tokenizer(tokenizer_params.tokenizer_model)
+    tokenizer.enable_padding(direction="right",
+                             pad_id=0,
+                             pad_type_id=0,
+                             pad_token=tokenizer_params.pad_token,
+                             pad_to_multiple_of=8)
+    tokenizer.normalizer = tokenizer_params.normalizer  # type: ignore
+    tokenizer.pre_tokenizer = tokenizer_params.pre_tokenizer  # type: ignore
+    tokenizer.decoder = tokenizer_params.decoder  # type: ignore
```

Comment:
type ignore?

Reply:
Is this because your are assigning attributes to tokenizer that are not in its public API?

Reply:
Correct! See this: https://github.com/mosaicml/composer/pull/650#discussion_r821361537

Is there a better way to do this? Happy to implement it if so

Reply:
It seems like that the underlying `.pyi` file has incorrect annotations. Let's paste in the error message here.

Reply:
Done!